src/main/java/com/limechain/network/ConnectionManager.java

```
@@ -23,23 +29,95 @@ public static ConnectionManager getInstance() {
+            }
+            protocolStreams.setInitiator(stream);
+        } else {
+            if (protocolStreams.getResponder() != null) {
+                stream.close();
+                return;
+            }
+            protocolStreams.setResponder(stream);
+        }
+    }
+
+    private PeerInfo addNewPeer(PeerId peerId) {
+        PeerInfo peerInfo = new PeerInfo();
         peers.put(peerId, peerInfo);
+        return peerInfo;
     }
 
-    public PeerInfo getPeerInfo(PeerId peerId) {
-       return peers.get(peerId);
+    public void closeGrandpaStream(Stream stream) {
+        PeerInfo peerInfo = peers.get(stream.remotePeerId());
+        if (peerInfo == null) {
+            return;
+        }
+
+        if (stream.isInitiator()) {
+            peerInfo.getGrandpaStreams().setInitiator(null);
+        } else {
+            peerInfo.getGrandpaStreams().setResponder(null);
+        }
+    }
+
+    public void closeBlockAnnounceStream(Stream stream) {
+        PeerInfo peerInfo = peers.get(stream.remotePeerId());
+        if (peerInfo == null) {
+            return;
+        }
+
+        if (stream.isInitiator()) {
+            peerInfo.getBlockAnnounceStreams().setInitiator(null);
+        } else {
+            peerInfo.getBlockAnnounceStreams().setResponder(null);
+        }
     }
 
     public void updatePeer(PeerId peerId, BlockAnnounceMessage blockAnnounceMessage) {
-        final var peer = peers.get(peerId);
+        PeerInfo peer = peers.get(peerId);
+        if (peer == null) {
+            return;
+        }
         updateLatestBlock(peer, blockAnnounceMessage.getHeader().getBlockNumber());
-
         if (blockAnnounceMessage.isBestBlock()) {
             updateBestBlock(peer, blockAnnounceMessage.getHeader());
         }
     }
 
+    public void updatePeer(PeerId peerId, BlockAnnounceHandshake blockAnnounceHandshake) {
+        PeerInfo peerInfo = peers.get(peerId);
+        if (peerInfo == null) {
+            return;
+        }
```

We should probably issue a warning log rather than silently return, as this case shouldn't happen